[docs] Add 3rd party libraries integration examples for Joy Input (#38541)
mui/material-ui: docs/data/joy/components/input/input.md

{{"demo": "InputVariables.js", "hideToolbar": true}}

> ## Third-party library integration

Reviewer: I think this header is more clear—"library integration" feels vague, when this is really about integrating specifically with formatting libs. 

I would also suggest making this an h3 (and headers below it h4s), and moving it up into the ## Customization section (maybe at the bottom of it).

```suggestion
### Third-party formatting
```

> ## Integration with 3rd party libraries

Reviewer 2: ```suggestion
## 3rd party libraries integration
```

Maybe? To remove one word?

To integrate Input with 3rd party libraries, create a wrapper component to map props with 3rd party component API.

> Then use the wrapper component as a value to `slotProps.input.component` of Joy UI Input.

Reviewer 2: ```suggestion
Then, use the wrapper component as a value to the `slotProps.input.component` slot of the Joy UI Input.
```

Is slot the correct name for what `slotProps.input.component` is?

Author: In my opinion, the slot is the input but `slotProps.input.component` is the specific property that will be passed to input slot.

Reviewer 2: Gotcha, so maybe this is more accurate?

```suggestion
Then, use the wrapper component as a value to the `slotProps.input.component` property of the Joy UI Input.
```

> To integrate the Input component with 3rd party libraries, create a wrapper component to map props with 3rd party component APIs.
> 
> Then, use the wrapper component as a value to the `slotProps.input.component` property of the Joy UI Input.

Reviewer: I had refer back to the original issue and read the Text Field doc linked there before I understood this. I think we need more detail. Here's how I'd write this:

```suggestion
The Input component can be integrated with third-party formatting libraries for more complex use cases. 

Create a wrapper component to map props with third-party component APIs.
Then pass that wrapper as a value to the `slotProps.input.component` property of the Joy UI Input.

The demos below illustrate how to do this with two popular libraries.
```

Reviewer: Also, it's not clear to me what's meant by "map props with third-party component APIs." Is there another way to say this?

Author: For example, Joy Input passes `ref` (refer to the HTML input) prop to the slot but the 3rd-party expect `inputRef`, so a wrapper component (the middle man) is needed to map the props:

Joy Input `ref` prop -> Wrapper (convert `ref` to `inputRef`) -> 3rd-party Input

Author: How about "Create an adapter component to get the props from the Input component and map them to the third-party component APIs."

### React imask

> [react-imask](https://github.com/uNmAnNeR/imaskjs) provides `IMaskInput` component for creating a wrapper component.

Reviewer 2: ```suggestion
[react-imask](https://github.com/uNmAnNeR/imaskjs) provides the vanilla JavaScript `IMaskInput` component enabling multiple ways to format text.
```

Similar to above!

> [react-imask](https://github.com/uNmAnNeR/imaskjs) provides the vanilla JavaScript `IMaskInput` component, which enables multiple ways to format text.

Reviewer: Maybe we should link here instead? I was a little confused when I landed on the `imaskjs` page. 

```suggestion
[react-imask](https://github.com/uNmAnNeR/imaskjs/tree/master/packages/react-imask) provides the `IMaskInput` component for complex formatting options.
```

{{"demo": "InputReactImask.js"}}

### React number format

> [react-number-format](https://github.com/s-yadav/react-number-format) provides `NumericFormat` component for creating a wrapper component.

Reviewer 2: ```suggestion
[react-number-format](https://github.com/s-yadav/react-number-format) provides the `NumericFormat` component enabling the formatting of texts that meet a specific number or string pattern.
```

Trying to convey what is the value behind this library in a more descriptive way. Pulled some words from their read.me.

> [react-number-format](https://github.com/s-yadav/react-number-format) provides the `NumericFormat` component, which enables the formatting of texts that meet a specific number or string pattern.

Reviewer: ```suggestion
[react-number-format](https://github.com/s-yadav/react-number-format) provides the `NumericFormat` component for enforcing text formatting that follows a specific number or string pattern.
```
